src/lib/serialization/schema.ts

```diff
@@ -0,0 +1,255 @@
+    T extends M.ContainerReflection ? JSONOutput.ContainerReflection :
+    T extends M.Reflection ? JSONOutput.Reflection :
+    // Types
+    T extends M.ArrayType ? JSONOutput.ArrayType :
+    T extends M.IntersectionType ? JSONOutput.IntersectionType :
+    T extends M.IntrinsicType ? JSONOutput.IntrinsicType :
+    T extends M.ReferenceType ? JSONOutput.ReferenceType :
+    T extends M.ReflectionType ? JSONOutput.ReflectionType :
+    T extends M.StringLiteralType ? JSONOutput.StringLiteralType :
+    T extends M.TupleType ? JSONOutput.TupleType :
+    T extends M.UnknownType ? JSONOutput.UnknownType :
+    T extends M.Type ? JSONOutput.SomeType : // Technically AbstractType, but the union is more useful
+    // Miscellaneous
+    T extends M.Comment ? JSONOutput.Comment :
+    T extends M.CommentTag ? JSONOutput.CommentTag :
+    T extends DecoratorWrapper ? JSONOutput.Decorator :
+    T extends SourceReferenceWrapper ? JSONOutput.SourceReference :
+    never;
+
+/**
+ * Contains interfaces which describe the JSON output. Each interface is related to a specific type of serializer.
+ *
+ * ## Plugins
+ * Plugins which modify the serialization process can use declaration merging
+ * to add custom properties to the exported interfaces.
+ * For example, if your custom serializer adds a property to all [[Reflection]] objects:
+ * ```ts
+ * declare module 'typedoc/dist/lib/serialization/schema' {
+ *     export namespace JSONOutput {
+ *         export interface AbstractReflection {
+ *             myCustomProp: boolean
+ *         }
+ *     }
+ * }
+ * ```
+ *
+ * If a plugin defines a new Model type, [[ModelToObject]] will not pick up the serializer type.
+ * To fix this, use declaration merging to augment the [[Serializer]] class.
+ * ```ts
+ * declare module 'typedoc/dist/lib/serialization/serializer' {
+ *     export interface Serializer {
+ *         toObject(value: CustomModel, obj?: Partial<CustomModel>): CustomOutput
+ *     }
+ * }
+ * ```
+ *
+ * For documentation on the JSON output properties, view the corresponding model.
+ */
+export namespace JSONOutput {
+
+    // Reflections
+
+    export interface ReflectionGroup {
+        title: M.ReflectionGroup['title'];
+        kind: M.ReflectionGroup['kind'];
+        children?: M.ReflectionGroup['children'][number]['id'][];
+        categories?: ModelToObject<M.ReflectionGroup['categories']>;
+    }
+
+    export interface ReflectionCategory {
+        title: M.ReflectionCategory['title'];
+        children?: M.ReflectionCategory['children'][number]['id'][];
+    }
+
+    export interface SignatureReflection extends Reflection {
+        type?: ModelToObject<M.SignatureReflection['implementationOf']>;
+        overwrites?: ModelToObject<M.SignatureReflection['implementationOf']>;
+        inheritedFrom?: ModelToObject<M.SignatureReflection['implementationOf']>;
+        implementationOf?: ModelToObject<M.SignatureReflection['implementationOf']>;
```

Comment:
Many of these look like they could be simplified using a [`keyof` mapping](http://www.typescriptlang.org/play/index.html#src=enum%20Complexity%20%7B%0D%0A%20%20%20%20first%2C%0D%0A%20%20%20%20second%0D%0A%7D%0D%0A%0D%0Aclass%20A%20%7B%0D%0A%20%20%20%20foo%3F%3A%20string%3B%0D%0A%20%20%20%20bar%3A%20B%3B%0D%0A%7D%0D%0A%0D%0Aclass%20B%20%7B%0D%0A%20%20%20%20id%3A%20number%3B%0D%0A%20%20%20%20foo%3A%20boolean%3B%0D%0A%20%20%20%20baz%3A%20Complexity%3B%0D%0A%7D%0D%0A%0D%0Atype%20_ModelToObject%3CT%3E%20%3D%0D%0A%20%20%20%20T%20extends%20B%20%3F%20BSerialized%20%3A%0D%0A%20%20%20%20T%20extends%20Complexity%20%3F%20string%20%3A%0D%0A%20%20%20%20T%20extends%20string%20%3F%20string%20%3A%0D%0A%20%20%20%20T%20extends%20number%20%3F%20number%20%3A%0D%0A%20%20%20%20any%3B%0D%0A%0D%0A%0D%0Atype%20SimpleSerializer%3CT%3E%20%3D%20%7B%20%5BP%20in%20keyof%20T%5D%3A%20_ModelToObject%3CT%5BP%5D%3E%20%7D%3B%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20ASerialized%20%7B%0D%0A%20%20%20%20foo%3A%20string%3B%0D%0A%20%20%20%20bar%3A%20BSerialized%0D%0A%7D%0D%0A%0D%0Atype%20ASerializedAuto%20%3D%20SimpleSerializer%3CA%3E%3B%0D%0A%0D%0A%0D%0A%0D%0Ainterface%20BSerialized%20%7B%0D%0A%20%20%20%20foo%3A%20number%3B%0D%0A%20%20%20%20baz%3A%20string%3B%0D%0A%7D).

Something like
```
type SignatureReflection = SimpleSerialization<Reflection>
```

This would help avoid issues when refactoring property names.

Reply:
How did I miss the names here... I was avoiding `keyof` since for some reflections, not all properties are exported (notably, `parent` is absent as JSON can't include circular data structures)

Reply:
Also, while it is certainly possible to use `Exclude` on the `keyof` result, this results in types which are displayed in a much less user friendly manner. The redundancy hurts, but I think it is probably worth it for types that can be easily explored.

Reply:
If we have common patterns (like `parent`) there should be a way to exclude that.

Reply:
I gave this a shot, looking into preventing specific properties from being serialized and I don't think it is the right way to go about this.

1. The list changes for (almost) every serialized reflection.
1. As soon as a new property is added to a reflection, the serializer types have to be updated, even though nothing changed there.

Instead I introduced a helper `S` local to this file that handles basic cases. It behaves similarly to `Pick`, but passes properties through `ModelToObject` if necessary. Thoughts?